packages/connect/src/device/DeviceCurrentSession.ts

```diff
@@ -0,0 +1,306 @@
+        const { signal } = this.abortController;
+        const abortPromise = new Promise<ReturnType<typeof fail>>(resolve =>
+            signal.addEventListener('abort', () => resolve(error(signal.reason))),
+        );
+        const callPromise = this.callLoop(type, msg, abortPromise);
+        this.callPromise = callPromise;
+        const response = await callPromise;
+        this.callPromise = undefined;
+        this.abortController = undefined;
+
+        if (!response.success) throw response.error;
+
+        const { payload } = response;
+
+        if (!(Array.isArray(resType) ? resType : resType.split('|')).includes(payload.type)) {
+            // handle possible race condition
+            // Bridge may have some unread message in buffer, read it
+            await scheduleAction(
+                abort =>
+                    this.transport.receive({
+                        session: this.session,
+                        protocol: this.protocol,
+                        signal: abort,
+                    }),
+                { timeout: 500 },
+            ).catch(() => {});
+
+            throw ERRORS.TypedError(
+                'Runtime',
+                `assertType: Response of unexpected type: ${payload.type}. Should be ${resType}`,
+            );
+        }
+
+        return payload;
+    }
+
+    /**
+     * Bridge version =< 2.0.28 throws "other call in progress" error.
+     * as workaround takeover transportSession (acquire) before sending Cancel, this will resolve previous pending call.
+     */
+    private needCancelWorkaround() {
+        return (
+            this.transport.name === 'BridgeTransport' &&
+            !versionUtils.isNewer(this.transport.version, '2.0.28')
+        );
+    }
+
+    private async callLoop<T extends Messages.MessageKey>(
+        type: T,
+        msg: Messages.MessagePayload<T>,
+        abortPromise: Promise<ReturnType<typeof fail>>,
+    ): Promise<ReturnType<typeof success<Messages.MessageResponse>> | ReturnType<typeof fail>> {
+        let [name, data] = [type, msg];
+        const { protocol, session } = this;
+        let pinUnlocked = false;
+
+        while (true) {
+            if (this.disposed) return disposed();
+
+            const callPromise = this.call({ session, name, data, protocol });
+
+            const abortedDuringCall = await Promise.race([
+                callPromise.then(() => false),
+                abortPromise.then(() => true),
+            ]);
+
+            if (name === 'ButtonAck' && abortedDuringCall && !this.disposed) {
+                if (this.needCancelWorkaround()) {
+                    try {
+                        // UI_EVENT is send right before ButtonAck, make sure that ButtonAck is sent
+                        await resolveAfter(1);
+                        await this.device.acquire();
+                        await this.device.getCurrentSession().cancelCall(false);
+                        await this.device.release();
```

Comment:
interesting that only ButtonRequest uses this workaround. I presume it was like this also before. There are couple of other "Something-Acks"

Reply:
Yeah, it was like this only for `ButtonAck`. The difference is that only `ButtonAck` expects cancel during pending `transport.call`, the other ones just expect it while they wait for data from UI promise.

Reply:
But of course, there are still places where `cancelCall` is called from outside `DeviceCurrentSession` so it can basically cancel any pending call and I'd be delighted to unify it with this case, instead of differentiating between various types.